src/components/entity/ha-entity-state-picker.ts

```diff
@@ -37,30 +50,90 @@ class HaEntityStatePicker extends LitElement {
 
   @property({ type: Boolean }) private _opened = false;
 
+  @state() private _stateHistory;
+
+  @state() private _fetching = false;
+
+  @state() private _discardFetch = false;
+
   @query("ha-combo-box", true) private _comboBox!: HaComboBox;
 
   protected shouldUpdate(changedProps: PropertyValues) {
     return !(!changedProps.has("_opened") && this._opened);
   }
 
+  private async _getStateHistory(): Promise<void> {
+    if (this._fetching) {
+      return;
+    }
+    this._fetching = true;
+    this._discardFetch = false;
+    try {
+      const endTime = new Date();
+      const startTime = new Date(endTime.getTime() - 60 * 60 * 24 * 1000 * 3); // 3 days
+      const history = {
+        ...(await fetchDateWS(this.hass, startTime, endTime, [this.entityId!])),
+      };
+      if (!this._discardFetch) {
+        this._stateHistory = history[this.entityId!].map((key) => key.s);
+        this._stateHistory = [...new Set(this._stateHistory.reverse())].slice(
+          0,
+          9
+        );
+      }
+    } finally {
+      this._fetching = false;
+      this._discardFetch = false;
+    }
+  }
+
   protected updated(changedProps: PropertyValues) {
+    if (
+      this.entityId &&
+      (changedProps.has("entityId") ||
+        (changedProps.has("_discardFetch") &&
+          changedProps.get("_discardFetch") &&
+          !this._discardFetch))
+    ) {
+      this._discardFetch = this._fetching; // if entity changes while fetching, discard the result, and then refetch when done
+      this._stateHistory = [];
+      if (DYNAMIC_STATE_DOMAINS.includes(computeDomain(this.entityId))) {
+        this._getStateHistory();
+      }
+    }
+
     if (changedProps.has("_opened") && this._opened) {
-      const state = this.entityId ? this.hass.states[this.entityId] : undefined;
-      (this._comboBox as any).items =
-        this.entityId && state
-          ? getStates(state, this.attribute).map((key) => ({
+      const hassState = this.entityId
+        ? this.hass.states[this.entityId]
+        : undefined;
+      const items =
+        this.entityId && hassState
+          ? getStates(hassState, this.attribute).map((key) => ({
               value: key,
               label: !this.attribute
                 ? computeStateDisplay(
                     this.hass.localize,
-                    state,
+                    hassState,
```

Comment:
I did not want to have to rename this variable, but one of the eslints was angry that I was trying to import "state", and also had an existing const named "state", I think it was getting confused.

Reply:
We use `entityState` usually now